internal/controller/dnsrecord_controller.go

```diff
@@ -106,40 +107,48 @@ func (r *DNSRecordReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
-		setDNSRecordCondition(dnsRecord, string(v1alpha1.ConditionTypeReady), metav1.ConditionFalse, reason, message)
-		return r.updateStatus(ctx, previous, dnsRecord, false, err)
-	}
-
+	// Update the logger with appropriate record/zone metadata
 	logger = log.FromContext(ctx).
+		WithValues("rootHost", dnsRecord.Spec.RootHost).
 		WithValues("ownerID", dnsRecord.Status.OwnerID).
-		WithValues("zoneID", managedZone.Status.ID)
+		WithValues("zoneID", dnsRecord.Status.ZoneID).
+		WithValues("zoneDomainName", dnsRecord.Status.ZoneDomainName)
 	ctx = log.IntoContext(ctx, logger)
 
 	if dnsRecord.DeletionTimestamp != nil && !dnsRecord.DeletionTimestamp.IsZero() {
 		logger.Info("Deleting DNSRecord")
-		if err = r.ReconcileHealthChecks(ctx, dnsRecord, managedZone); client.IgnoreNotFound(err) != nil {
-			return ctrl.Result{}, err
-		}
-		hadChanges, err := r.deleteRecord(ctx, dnsRecord, managedZone)
-		if err != nil {
-			logger.Error(err, "Failed to delete DNSRecord")
-			return ctrl.Result{}, err
-		}
-		// if hadChanges - the deleteRecord has successfully applied changes
-		// in this case we need to queue for validation to ensure DNS Provider retained changes
-		// before removing finalizer and deleting the DNS Record CR
-		if hadChanges {
-			return ctrl.Result{RequeueAfter: randomizedValidationRequeue}, nil
+		if dnsRecord.Status.ZoneID != "" && dnsRecord.Status.ZoneDomainName != "" {
+			// Create a dns provider with config calculated for the current dns record status (Last successful)
+			dnsProvider, err := r.ProviderFactory.ProviderFor(ctx, dnsRecord, provider.Config{
+				ZoneIDFilter:   externaldnsprovider.NewZoneIDFilter([]string{dnsRecord.Status.ZoneID}),
+				DomainFilter:   externaldnsendpoint.NewDomainFilter([]string{dnsRecord.Status.ZoneDomainName}),
+				ZoneTypeFilter: externaldnsprovider.NewZoneTypeFilter(""),
+			})
+			if err != nil {
+				logger.Error(err, "Failed to load DNS Provider")
+				reason := "DNSProviderError"
+				message := fmt.Sprintf("The dns provider could not be loaded: %v", err)
+				setDNSRecordCondition(dnsRecord, string(v1alpha1.ConditionTypeReady), metav1.ConditionFalse, reason, message)
+				return r.updateStatus(ctx, previous, dnsRecord, false, err)
+			}
+
+			if err = r.ReconcileHealthChecks(ctx, dnsRecord); client.IgnoreNotFound(err) != nil {
+				return ctrl.Result{}, err
+			}
+			hadChanges, err := r.deleteRecord(ctx, dnsRecord, dnsProvider)
+			if err != nil {
+				logger.Error(err, "Failed to delete DNSRecord")
+				return ctrl.Result{}, err
+			}
+			// if hadChanges - the deleteRecord has successfully applied changes
+			// in this case we need to queue for validation to ensure DNS Provider retained changes
+			// before removing finalizer and deleting the DNS Record CR
+			if hadChanges {
+				return ctrl.Result{RequeueAfter: randomizedValidationRequeue}, nil
+			}
+		} else {
+			logger.Info(fmt.Sprintf("unable to perform cleanup of dns provider for zone status id: '%s', dominaName: '%s'",
```

Comment:
is it possible to give more info than this? The way I read it, this can happen if we failed to set a zid or domain inside the zone. As this is a delete also makes me wonder if this could happen. I guess if a record was created and there was a bad credential and then you tried to delete it?

Reply:
Yes, if the record is created and for whatever reason it isn't able to assign a zone e.g. invalid credentials, no zone that matches the root host etc... we shouldn't try and cleanup when the record is deleted since it won't have been able to create anything in the first place.